Source/Error.swift

```diff
@@ -24,428 +24,540 @@
+             .outputStreamCreationFailed(let url), .outputStreamFileAlreadyExists(let url), .outputStreamURLInvalid(let url),
+             .bodyPartFileNotReachableWithError(let url, _), .bodyPartFileSizeQueryFailedWithError(let url, _):
             return url
         default:
             return nil
         }
     }
-    
-    var outputStream: OutputStream? {
-        guard case let .writeFailed(outputStream) = self else { return nil }
-        
-        return outputStream
+
+    var underlyingError: Error? {
+        switch self{
+        case .bodyPartFileNotReachableWithError(_, let error), .bodyPartFileSizeQueryFailedWithError(_, let error),
+             .outputStreamWriteFailed(let error), .inputStreamReadFailed(let error):
+            return error
+        default:
+            return nil
+        }
     }
-    
+
 }
 
 extension AFError.ValidationFailureReason {
-    
+
     var acceptableContentTypes: [String]? {
-        guard case let .contentTypeMismatch(acceptableContentTypes, _) = self else { return nil }
-        
-        return acceptableContentTypes
-    }
-    
-    var actualContentType: String? {
-        guard case let .contentTypeMismatch(_, actualContentType) = self else { return nil }
-        
-        return actualContentType
-    }
-    
-    var statusCode: Int? {
-        guard case let .unacceptableStatusCode(statusCode) = self else { return nil }
-        
-        return statusCode
-    }
-    
+        switch self {
+        case .missingContentType(let types), .unacceptableContentType(let types, _):
+            return types
+        default:
+            return nil
+        }
+    }
+
+    var responseContentType: String? {
+        switch self {
+        case .unacceptableContentType(_, let reponseType):
+            return reponseType
+        default:
+            return nil
+        }
+    }
+
+    var responseCode: Int? {
+        switch self {
+        case .unacceptableStatusCode(let code):
+            return code
+        default:
+            return nil
+        }
+    }
+
 }
 
 extension AFError.SerializationFailureReason {
-    
-    var stringEncoding: String.Encoding? {
-        guard case let .invalidStringEncoding(encoding) = self else { return nil }
-        
-        return encoding
+
+    var failedStringEncoding: String.Encoding? {
+        switch self {
+        case .stringSerializationFailed(let encoding):
+            return encoding
+        default:
+            return nil
+        }
+    }
+
+    var underlyingError: Error? {
+        switch self {
+        case .jsonSerializationFailed(let error), .propertyListSerializationFailed(let error):
+            return error
+        default:
+            return nil
+        }
     }
-    
+
 }
 
 // MARK: Error Descriptions
 
 extension AFError: LocalizedError {
-    
+
     public var errorDescription: String? {
```

Comment:
The docstring is provided by the protocol. If you option click this function you'll see it. Unless you want us to provide our own?

![screen shot 2016-08-28 at 4 35 56 pm](https://cloud.githubusercontent.com/assets/51020/18036613/8e523dae-6d3d-11e6-9190-44aa773028d1.png)